Add NVD3's bullet chart

darabos opened the pull request:
Hi Superset people,
[We](http://www.lynxanalytics.com) are considering adopting Superset for some projects. It looks very nice, thanks for all the work. To get to know it better before jumping in with both feet I wanted to try adding a new chart type that we would want to use anyway. [NVD3's bullet chart](http://nvd3.org/examples/bullet.html) looked nice and simple. But on a closer look it turned out to have quite a few parameters!

Here's a draft implementation for this to get your thoughts on it. It seems to work in "ExploreV2" mode. I'm not sure about the V1/V2 distinction. A couple issues:
 - It displays an error until you set a range.
 - The measure line's color does not get set for some reason.
 - Would be nice to provide a way to set the ranges/markers from metrics as well.
 - The bullet chart is stretched to fill the height of the panel. It looks pretty bad until you resize the panel.
 - I set `is_timeseries = False` but still there are time settings at the bottom of the explore UI. Is that normal?

Thanks for any feedback!

darabos committed:
Add NVD3's bullet chart.

mistercrunch commented:
please add a thumbnail of the right size (same widht/height as other thumbs) for the visualization dropdown, and add an entry in the docs (gallery.rst).  Also please fix the 2 minor codeclimate issues

darabos committed:
Add empty lines before nested function definitions.

darabos committed:
Add thumbnail for bullet chart.

darabos committed:
Add bullet chart to gallery.rst.

darabos commented:
Thanks for the quick response!

> please add a thumbnail of the right size (same widht/height as other thumbs) for the visualization dropdown, and add an entry in the docs (gallery.rst). Also please fix the 2 minor codeclimate issues

Done.

ascott commented:
thx @darabos!

> It displays an error until you set a range.
> The measure line's color does not get set for some reason.
> Would be nice to provide a way to set the ranges/markers from metrics as well.
> The bullet chart is stretched to fill the height of the panel. It looks pretty bad until you resize the panel.

^^^ we should fix these things before we ship 

> I set is_timeseries = False but still there are time settings at the bottom of the explore UI. Is that normal?

^^^ this sounds like a bug, filed an issue here: https://github.com/airbnb/superset/issues/1777

ascott commented:
also, V1 will be deprecated soon, so this only needs to work in V2. we are doing some beta testing now and will be fully moving over to V2 in the next couple weeks.

ascott commented on superset/assets/javascripts/explorev2/stores/store.js:
```diff
@@ -407,6 +407,21 @@ export const visTypes = {
   },
 
+  bullet: {
+    label: 'Bullet Chart',
```
the `bullet` object needs to contain `requiresTime: false,` so the time controls are not rendered.

darabos replied:
> the `bullet` object needs to contain `requiresTime: false`, so the time controls are not rendered.

Done. I still get the time controls, but perhaps it's due to testing with 0.14.1 instead of `master`. (Sorry.)

mistercrunch commented:
Trying to answer your questions:
* v1 will be deprecated as soon as v2 is 100% working, should be a few weeks
* v2 will allow for arbitrarily complex controls 
* `is_timeseries=false` still allows for time filtering, but should hide `Granularity`
* you can add a `height` parameter if you don't want height to fill the box, consider adding an option to fill the container, perhaps when height=0

I don't understand "Would be nice to provide a way to set the ranges/markers from metrics as well."

ascott commented on superset/forms.py:
```diff
@@ -981,6 +981,36 @@ def __init__(self, viz):
                 ],
                 "description": _("The color for points and clusters in RGB")
             }),
+            'ranges': (TextField, {
```
we shouldn't need to add these here for V2

ascott commented on superset/viz.py:
```diff
@@ -896,6 +896,69 @@ def get_data(self):
+
+    """Based on the NVD3 bullet chart"""
+
+    viz_type = "bullet"
```
lines 903-914 can be deleted, not needed for V2

darabos replied:
> lines 903-914 can be deleted, not needed for V2

`viz_type` is still used in the last line of `viz.py` for registration. I cannot get it working if I remove the other lines either. I only know how to turn on V2 for the explore UI. But it does not affect the dashboard UI I think.

Is it okay if I keep the V1 lines for now? I assume they will be mass-deleted at some point. But it's easier for me to debug things in an unknown codebase if I can just follow the examples.

darabos committed:
Add "requiresTime: false", fix indentation.

darabos committed:
Avoid scaling bullet chart vertically.

darabos committed:
Use a default if no range is specified.

darabos committed:
Fix coloring of bullet chart.

darabos commented:
> we should fix these things before we ship

> > It displays an error until you set a range.

Fixed. If no range is set, I set the range to 0-110%.

> > The measure line's color does not get set for some reason.

This was due to the `chart.color((d) => category21(d[colorKey]));` line in `nvd3_vis.js`. I don't have a deeper understanding of this, but I skip this line for the bullet chart now.

> > Would be nice to provide a way to set the ranges/markers from metrics as well.

By this I mean maybe you have two metrics, `SALES` and `SALES_TARGET`. Then you might want to display `SALES` as the progress bar, but put a triangle marker at `SALES_TARGET` instead of a constant. In my current version you can only specify constants.

One easy solution would be to treat non-numeric values in the `markers`/`marker_lines`/`ranges` parameters as metric names. (Maybe also for the labels.) I think I could implement this, but I see no precedent for it. Metrics are always picked from dropdowns, while here the user would have to type the metric name.

What do you think? I guess it would not actively _hurt_ anyone, so I would be inclined to implement this. Of course if you have a better idea, that would be best.

> > The bullet chart is stretched to fill the height of the panel. It looks pretty bad until you resize the panel.

Fixed. Its height is limited to 50px now.

darabos committed:
Merge branch 'master' of github.com:airbnb/superset into darabos-bullet-chart

darabos commented:
I've merged it to current `master`. Seems to still work! :smile:

mistercrunch commented:
LGTM
